### DISCUSSION TITLE
encoding of json::exception::what()

### DISCUSSION THREAD

**Discussion by razaqq** [nlohmann/json#2824] [Q&A] [answered]
So im trying to work with the error messages provided by json::exception::what()

For example i purposefully tried parsing jibberish like `xyz` with this

```c++
class sax_no_exception : public nlohmann::detail::json_sax_dom_parser<json>
{
public:
	explicit sax_no_exception(json& j) : nlohmann::detail::json_sax_dom_parser<json>(j, false) {}

	static bool parse_error(std::size_t position, const std::string& last_token, const json::exception& ex)
	{
		std::cout << ex.what() << std::endl;
		return false;
	}
};
```

but instead of getting an error message, i ended up with this:
`¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦╗▀Ôx¦╔`

I tried multiple different terminals, i tried `SetConsoleOutputCP( CP_UTF8 );` and i tried dumping to a file and trying to read it as UTF-8, ANSI, etc. Unfortunately all without luck.

I get the same result when using this:
```c++
#define JSON_THROW_USER(exception)                           \
    {std::clog << "Error in " << __FILE__ << ":" << __LINE__ \
               << " (function " << __FUNCTION__ << ") - "    \
               << (exception).what() << std::endl;           \
     std::abort();}
```

When looking in memory at the location of `ex.m.what()` i get this
![grafik](https://user-images.githubusercontent.com/5948878/122419990-9f015300-cf8b-11eb-94d9-8c98fb580532.png)

I have exceptions turned off. Using clang with cmake, not that that should matter
Any ideas?

**Comment by nlohmann** [comment, marked as answer]
I'll check. Can you please elaborate what you mean with "I have exceptions turned off."?

**Reply by razaqq** [reply]
compiler flag `-fno-exceptions` (works for gcc and clang)

Also i forgot to mention im using version `3.9.1` (latest) of your awesome library!

I will try to create a small testcase for you in a second to not waste your time.

EDIT: here it is: https://github.com/razaqq/json_bug_testcase

**Reply by nlohmann** [reply]
Thanks! I cannot reproduce the issue on macOS though.

Output is

```
[json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - invalid literal; last read: 'x'
Error while parsing JSON
```

Note I had to use the latest `develop` version as there has been a bug as not all `throw`s were replaced by `JSON_THROW`.

**Reply by razaqq** [reply]
As i said im on windows with clang and microsofts STL version `16.10`

```
λ clang --version
clang version 12.0.0
Target: x86_64-pc-windows-msvc
Thread model: posix
```

I will test with the develop version now and see if there is a different result.

**Reply by razaqq** [reply]
Okay so it still happens even after bumping the version of json to `develop`.

I have created a Github Actions CI that builds on windows with cmake, clang and microsoft STL and the exact same behavior can be seen there.

[Link to CI](https://github.com/razaqq/json_bug_testcase/runs/2925131820), check the output of the run step.

This means this issue is not only related to my machine, i hope this helps you reproduce this.

**Reply by razaqq** [reply]
Any updates on this?

**Reply by gregmarr** [reply]
Works fine on compiler explorer with Clang 12.0.1 too:  https://godbolt.org/z/KWYe68q9h
That's with trunk as it doesn't have 3.9.1 installed.  Same with develop.

**Reply by gregmarr** [reply]
I notice this is a debug build.  Have you tried a release build to see if the behavior is the same?

**Reply by gregmarr** [reply]
Another thing to try would be to remove `-fno-exceptions` and add `JSON_NOEXCEPTION` to see if it's an interaction with the exception flag and MS STL.

**Reply by razaqq** [reply]
i get a different behavior in release mode, but still no error message
```
`TüJY☻
Error while parsing JSON
```

i also tried compiling with clang-cl without `-fno-exceptions`, but with `add_compile_definitions(JSON_NOEXCEPTION)` but with the same result

**Reply by nlohmann** [reply]
I created a branch https://github.com/nlohmann/json/pull/2898 with a regression test and a new test suite that is executed with disabled exceptions. I'm not sure if the flags and defines I use for MSVC are correct (https://github.com/nlohmann/json/pull/2898/files#diff-33394812ba204689144fd2f80832db83853ba1cb32403edb4e15fe4893e675fd), but at least I can reproduce the issue in the CI with MSVC.

I'm not sure if the library is doing anything wrong - if I had to guess, the behavior of classes derived from `std::exception` or members like `std::runtime_error` may differ when exceptions have been switched off.

**Reply by nlohmann** [reply]
Here are the failing tests: https://github.com/nlohmann/json/actions/runs/1087268990 - note that Clang and Clang-CL seem to work, whereas all MSVC tests fail.

**Reply by gregmarr** [reply]
@nlohmann That's interesting, because this report is for issues with Clang.  I guess it's probably because it's Microsoft's STL.

**Reply by nlohmann** [reply]
I'll try to find a smaller example which (hopefully) does not involve the whole library, but rather the exception class.

**Reply by razaqq** [reply]
Im glad you guys are able to reproduce this. 

Yeah it could end up being some clang <-> microsoft stl issue...

**Reply by nlohmann** [reply]
I now have a minimal example that is independent of the library. The following code prints garbage when compiled with MSVC with exceptions disabled, but is fine with other compilers (Clang, GCC):

```cpp
#include <exception> // exception
#include <stdexcept> // runtime_error
#include <string> // string
#include <iostream> // cout

namespace nlohmann
{
namespace detail2
{

class exception : public std::exception
{
  public:
    const char* what() const noexcept override
    {
        return m.what();
    }

  protected:
    exception(const char* what_arg) : m(what_arg) {}

  private:
    std::runtime_error m;
};

class parse_error : public exception
{
  public:
    static parse_error create(const std::string& what_arg)
    {
        std::string w = "[json.exception.parse_error] " + what_arg;
        return parse_error(w.c_str());
    }

  private:
    parse_error(const char* what_arg) : exception(what_arg) {}
};

}  // namespace detail2
}  // namespace nlohmann


int main()
{
    auto error = nlohmann::detail2::parse_error::create("foo");
    std::cout << error.what() << std::endl;
}
```

I am using these CMake flags:

```cmake
target_compile_options(test-disabled_exceptions PUBLIC /EH)
target_compile_definitions(test-disabled_exceptions PUBLIC _HAS_EXCEPTIONS=0)
```

Can someone using Windows reproduce this?

**Reply by gregmarr** [reply]
I debugged into this and MS STL does not make a copy of the string passed to `explicit exception(char const* const _Message) noexcept` when `_HAS_EXCEPTIONS=0` is set.

**Reply by nlohmann** [reply]
I assume this is a bug, right?

**Reply by gregmarr** [reply]
According to my understanding of the standard, the exception is supposed to copy the data, and it does for the regular case.

However, the standard doesn't provide implementations the ability to turn off exceptions, so this is entirely non-standard behavior, and thus you can't really use the standard to say what it is or isn't supposed to do.  I can see a case for considering this a bug as it is definitely a variance in class behavior between exceptions enabled and not.

I don't imagine that they expected anyone to actually use `std::exception` with exceptions turned off.  The compiler actually warns about using `try` while exceptions are disabled.  I would be surprised if they changed the behavior of this any time soon.  So I'm not sure that this is something that the library can do much about except saying that if you have `_HAS_EXCEPTIONS=0` set in your compiler settings and you're using MS STL, then `std::exception::what()` can not be used.

**Reply by nlohmann** [reply]
Thanks for checking. I also had a look:

https://github.com/microsoft/STL/blob/472161105d596192194d4715ccad307c6c163b4a/stl/inc/exception#L90

Would it make sense to open an issue at https://github.com/microsoft/STL for this?

In the meantime, I will add a note to the documentation.

**Reply by gregmarr** [reply]
Yes, that's the code that we hit in the _HAS_EXCEPTIONS=0 case.  I think since it's available on GitHub, filing an issue would be appropriate.

**Reply by nlohmann** [reply]
I filed a bug: https://github.com/microsoft/STL/issues/2114

**Reply by nlohmann** [reply]
It is indeed a bug in Microsoft's STL. I think there is nothing we can do at this point until it is fixed.

**Reply by razaqq** [reply]
thanks for your effort of investigating this! much appreciated

**Reply by gregmarr** [reply]
You're welcome.
